### ISSUE TITLE
Support .js config files in "type": "module" projects?

### ISSUE THREAD

**Issue by michael42** [antonk52/lilconfig#29]
Hi there,

would you consider supporting `.js` files inside projects with `"type": "module"` set? Otherwise is pretty hard to provide config files sometimes, because not every user of `lilconfig` actually includes `.cjs` as an allowed file extension.

Currently I'm using the following patch, which should work for CommonJS and ECMAScript Modules since [Node 12](https://nodejs.org/dist/latest-v12.x/docs/api/esm.html#esm_import_expressions), so it would require a semver-major bump (or a fallback for Node 10):

```diff
diff --git a/dist/index.js b/dist/index.js
index a153607e305eb6d40b735aa176eee388b8172dc9..497efc7dc60165b306f46211fb7ddd811b51ef72 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -28,7 +28,7 @@ function getSearchPaths(startDir, stopDir) {
     }, { searchPlaces: [], passedStopDir: false }).searchPlaces;
 }
 exports.defaultLoaders = Object.freeze({
-    '.js': require,
+    '.js': id => import(id).then(mod => mod.default),
     '.json': require,
     '.cjs': require,
     noExt(_, content) {
@@ -123,7 +123,7 @@ function lilconfig(name, options) {
                 }
                 else {
                     validateLoader(loader, loaderKey);
-                    result.config = loader(filepath, content);
+                    result.config = await loader(filepath, content);
                 }
                 result.filepath = filepath;
                 break;
```

**Comment by antonk52** [comment]
Hi and thank you for the issue.

Since this package attempts to mimic cosmiconfig, I would rather to wait for them to make a major release with support for ESM native node modules and drop nodejs v10. From my understanding it is not yet implemented as nodejs does not offer an API to clear cache for imported modules like `require.cache` which is a blocker for cosmiconfig.

From your diff I can see that there is currently an issue with lilconfig as loaders can be async for `lilconfig` function. You can submit a PR to add the two missing `await`s and I am happy to merge it and publish a new version that would allow you to provide an ESM loader instead of patching a dependency, ie

```js
const loadESMDefault = p => import(p).then(x => x.default)

lilconfig('myapp', {
  loaders: {
    '.js': loadESMDefault,
    '.mjs': loadESMDefault,
  }
}).load('./path/to/esm/file.js') // Promise<config>
```

Await is missing [here](https://github.com/antonk52/lilconfig/blob/029f357962865ef66037728e15eb9fe4bbf5a5ed/src/index.ts#L214) and [here](https://github.com/antonk52/lilconfig/blob/029f357962865ef66037728e15eb9fe4bbf5a5ed/src/index.ts#L234).

Let me know what you think about this.

**Comment by michael42** [comment]
Thanks for the fast response and great idea splitting the patch. I created a [pull request](https://github.com/antonk52/lilconfig/pull/30) adding the `await`s (and learned a bit about the API while writing a test for it).

**Comment by antonk52** [comment]
Nicely done. I've released a new version [v2.0.5](https://github.com/antonk52/lilconfig/releases/tag/v2.0.5) with your fix, this should make it easier for people to consume ESM configs.

**Comment by marekdedic** [comment]
Hi, it seems to me that osmiconfig now supports this looking at the changelog for version 8.2.0. Could we please reopen this? Happy to send a PR if necessary...

**Comment by antonk52** [comment]
Hi @marekdedic this sounds reasonable to me. Happy to review and merge your PR. I’ve reopened the issue.

**Comment by antonk52** [comment]
closed by #46
